src/main/java/io/vertx/core/impl/VertxImpl.java

```diff
@@ -870,17 +871,18 @@ private class InternalTimerHandler implements Handler<Void>, Closeable, Runnable
     private final long timerID;
     private final ContextInternal context;
     private final java.util.concurrent.Future<?> future;
+    private final AtomicBoolean disposed = new AtomicBoolean();
 
-    InternalTimerHandler(long timerID, Handler<Long> runnable, boolean periodic, long delay, ContextInternal context) {
+    InternalTimerHandler(long timerID, Handler<Long> runnable, boolean periodic, long delay, TimeUnit timeUnit, ContextInternal context) {
       this.context = context;
       this.timerID = timerID;
       this.handler = runnable;
       this.periodic = periodic;
       EventLoop el = context.nettyEventLoop();
       if (periodic) {
```

Comment:
The proposed solution does ensure that the handler will fire regardless of what happens with the entry in the `timeouts` map. This resolves my concerns I had about my solution of how ti would impact cancelling of the timers. Nice!

But the root cause of the issue that the InternalTimerHandler might see a hasmap without the entry when it tries to  remove it remains in this implementation. 
In this case the handler would work as expected but we would be leaving garbage entries in the `timeouts` map that would never be removed, having a very subtle memory leak.
I would suggest to also move the setting of the netty scheduling outside the constructor in `scheduleTimeout()` after the timer has been added to `timeouts` like I have implemented in the second version of my PR: #4371

Reply:
good catch, I actually had local changes I forgot to commit, I forced pushed with all the changes.

PTAL